### PR TITLE
Zygote ext

### PR DESCRIPTION
**Pull request by vpuri3**
part of https://github.com/JuliaDiff/SparseDiffTools.jl/issues/215

ref https://pkgdocs.julialang.org/dev/creating-packages/#Conditional-loading-of-code-in-packages-(Extensions)

### CODE REVIEW

**Comment by vpuri3** [comment]
running CI with dummy change since tests were failing locally.

File: src/SparseDiffTools.jl [hunk 1]
```diff
@@ -68,16 +68,16 @@ Base.@pure __parameterless_type(T) = Base.typename(T).wrapper
...
 
-#if !isdefined(Base, :get_extension)
+if !isdefined(Base, :get_extension)
     using Requires
-#end
+end
 
 function __init__()
-    #@static if !isdefined(Base, :get_extension)
+    @static if !isdefined(Base, :get_extension)
```
Comment (ChrisRackauckas) [inline src/SparseDiffTools.jl:76]: Move this outside the init

File: ext/SparseDiffToolsZygote.jl [hunk 2]
```diff
@@ -1,3 +1,24 @@
+module SparseDiffToolsZygote
+
+import Zygote
+
+using SparseDiffTools
+using SparseDiffTools: DeviVecTag, FwdModeAutoDiffVecProd
+
+using ForwardDiff
```
Comment (ChrisRackauckas) [inline ext/SparseDiffToolsZygote.jl:8]:
```suggestion
using SparseDiffTools.ForwardDiff
```

File: ext/SparseDiffToolsZygote.jl [hunk 3]
```diff
@@ -1,3 +1,24 @@
...
+using ForwardDiff: Dual, Tag
+
+using SciMLOperators: FunctionOperator
```
Comment (ChrisRackauckas) [inline ext/SparseDiffToolsZygote.jl:11]:
```suggestion
using SparseDiffTools.SciMLOperators: FunctionOperator
```

File: ext/SparseDiffToolsZygote.jl [hunk 4]
```diff
@@ -1,3 +1,24 @@
...
+using Tricks: static_hasmethod
```
Comment (ChrisRackauckas) [inline ext/SparseDiffToolsZygote.jl:12]:
```suggestion
using SparseDiffTools.Tricks: static_hasmethod
```

File: ext/SparseDiffToolsZygote.jl [hunk 5]
```diff
@@ -1,3 +1,24 @@
+module SparseDiffToolsZygote
+
+import Zygote
+
+using SparseDiffTools
+using SparseDiffTools: DeviVecTag, FwdModeAutoDiffVecProd
```
Comment (ChrisRackauckas) [inline ext/SparseDiffToolsZygote.jl:6]: Needs to look like https://github.com/JuliaArrays/ArrayInterface.jl/blob/master/ext/ArrayInterfaceGPUArraysCoreExt.jl

**Comment by vpuri3** [comment]
Looks like package extension don't honor export statements. Rather, it is just 

```julia
julia> versioninfo()
Julia Version 1.9.0-rc1
Commit 3b2e0d8fbc1 (2023-03-07 07:51 UTC)

(a) pkg> st                                                                          
Status `~/.julia/dev/SparseDiffTools/a/Project.toml`
  [47a9eef4] SparseDiffTools v2.0.0 `..`
  [e88e6eb3] Zygote v0.6.56                                                          
                                          
julia> using SparseDiffTools, Zygote  
[ Info: Precompiling SparseDiffToolsZygote [25f82377-10ae-5fb8-bc0d-27fd31de4285]
                                          
julia> m = Base.get_extension(SparseDiffTools, :SparseDiffToolsZygote)
SparseDiffToolsZygote 

julia> m.ZygoteVecJac
ZygoteVecJac (generic function with 1 method)

julia> ZygoteVecJac
ERROR: UndefVarError: `ZygoteVecJac` not defined

julia> SparseDiffTools.ZygoteVecJac
ERROR: UndefVarError: `ZygoteVecJac` not defined
Stacktrace:
 [1] getproperty(x::Module, f::Symbol)
```

So I think we should define and export these symbols in the main package, and write method definitions in the extension.

File: Project.toml [hunk 6]
```diff
@@ -19,6 +19,12 @@ StaticArrays = "90137ffa-7385-5640-81b9-e52037218182"
 Tricks = "410a4b4d-49e4-4fbc-ab6d-cb71b17b3775"
 VertexSafeGraphs = "19fa3120-7c27-5ec5-8db8-b0b0aa330d6f"
 
+[weakdeps]
+Zygote = "e88e6eb3-aa80-5325-afca-941959d7151f"
+
+[extensions]
+SparseDiffToolsZygote = "Zygote"
```
[the remark this answers is not in the document]
Comment (vpuri3) [reply Project.toml:27]: compat added. already in extra.

**Comment by vpuri3** [comment]
@ChrisRackauckas can you merge https://github.com/JuliaDiff/SparseDiffTools.jl/pull/220 so we can test for 1x9 here?

**Comment by vpuri3** [comment]
rerunning CI

**Comment by vpuri3** [comment]
@ChrisRackauckas good to go

**Pull request merged**